qiskit/transpiler/passes/optimization/collect_linear_functions.py

```diff
@@ -0,0 +1,84 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Replace each sequence of CX and SWAP gates by a single LinearFunction gate."""
+
+from qiskit.circuit.library.standard_gates import CXGate, SwapGate
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.circuit import QuantumCircuit
+
+
+class CollectLinearFunctions(TransformationPass):
+    """Collect blocks of linear gates (CX and SWAP) and replace them by linear functions."""
+
+    @staticmethod
+    def _is_linear_gate(op):
+        return isinstance(op, (CXGate, SwapGate)) and op.condition is None
+
+    # Called when reached the end of the linear block (either the next gate
+    # is not linear, or no more nodes)
+    @staticmethod
+    def _finalize_processing_block(cur_nodes, cur_qubits, blocks):
+        # Collect only blocks comprising at least 2 gates.
+        # ToDo: possibly make the minimum number of gates as a parameter
+        if len(cur_nodes) >= 2:
+            cur_wire_pos_map = dict((qb, ix) for ix, qb in enumerate(cur_qubits))
```

Comment:
Since `cur_qubits` is a `set`, I'm not sure there will be a guaranteed order when `enumerate`-ing over it. Does the order of `ix` matter here?

Reply:
Thanks! I see your point. It should not matter though, since `cur_wire_pos_map` is also used when constructing an auxiliary linear quantum circuit representing the block, with that auxiliary quantum circuit passed to the constructor of `LinearFunction` (lines 74--80 in code).  That is, a different ordering in `wire_pos_map` would also result in a different but matching ordering in the `LinearFunction` itself. However. this makes me wonder if I maybe could simplify the code somehow.

Reply:
Thanks! I sort of understand the point that we can construct a `QuantumCircuit` from a list of `QuantumRegister`s and that we can add gates to this circuit by referring to registers directly. However, it seems that I still need to make a choice on how to order the qubits involved, and that I still need to construct the index mapping as it is needed for `replace_block_with_op`. Nevertheless, I will try to make the code more clear, and for reproducibility I will take the ordering compatible with the original ordering of qubits.